BUG: Add SparseArray.all

Licht-T opened the pull request:
- [x] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This is the part of https://github.com/pandas-dev/pandas/pull/17386.
`Block.where` uses `ndarray.all`, but there is no such method in `SparseArray`.
This makes that `all` evaluates `SparseArray([False, False, True])` as `True`.
https://github.com/pandas-dev/pandas/blob/master/pandas/core/internals.py#L1400

Licht-T committed:
TST: Add tests of SparseArray.all

jreback commented on pandas/core/sparse/array.py:
```diff
@@ -614,6 +614,24 @@ def fillna(self, value, downcast=None):
+        values = self.sp_values
+
+        if len(values) != len(self):
+            values = values.tolist()
```
is there a reason you are converting to a list? and not a dense array?

Licht-T replied:
@jreback For the performance reason. This is sparse and the converting to dense is not efficient.

jreback replied:
I am asking why you are converting ``.tolist()``, then appending, that is non-performant.

Licht-T replied:
Simply I wanted to use `np.all`. I got an idea not to use `tolist()`. I'll change the solution.

jreback reviewed (changes requested):
pls add a whatsnew note. do we have a specific issue for this? (other that the PR)?

jreback commented on pandas/core/sparse/array.py:
```diff
@@ -614,6 +614,24 @@ def fillna(self, value, downcast=None):
                                 fill_value=fill_value)
 
+    def all(self, axis=0, *args, **kwargs):
+        """
```
might as well add any as well.

Licht-T committed:
TST: Add tests of SparseArray.any

Licht-T committed:
DOC: Add SparseArray.all and SparseArray.any to whatsnew note

Licht-T commented:
@jreback Thanks for your review.
- Changed the `fill_value` check and now we are not using `tolist()`
- Also added `SparseArray.any`.
- There seems to be no specific issue for this.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -544,6 +544,7 @@ Sparse
 
 - Bug in ``SparseSeries`` raises ``AttributeError`` when a dictionary is passed in as data (:issue:`16905`)
 - Bug in :func:`SparseDataFrame.fillna` not filling all NaNs when frame was instantiated from SciPy sparse matrix (:issue:`16112`)
```
say that these are now implemented to handle SparseArray

jreback commented on pandas/core/sparse/array.py:
```diff
@@ -614,6 +614,40 @@ def fillna(self, value, downcast=None):
+
+        Returns
+        -------
+        all : bool
```
add a See Also to point to np.all

jreback reviewed (approved):
minor comment. ping on green

Licht-T commented:
@jreback Thank you. These are now fixed.

Licht-T commented:
@jreback I found the bug in this solution. When `fill_value = np.array([1,1,1])`, `np.all(fill_value)` returns True. I'll fix.

jreback commented:
lgtm. waiting for CI to go green again before merging (unrelated conda-ish issues)

jreback commented:
can you rebase

Licht-T committed:
BUG: Add SparseArray.any

Licht-T committed:
BUG: Add SparseArray.all

Licht-T committed:
Merge branch 'master' into add-sparsearray-all

Licht-T commented:
@jreback Rebased.

jreback commented:
thanks @Licht-T
